src/calculate/pass_4/rah.rs

```diff
@@ -0,0 +1,186 @@
+use crate::calculate::item::Object;
+use crate::calculate::pass_3::Cache;
+use crate::calculate::Ship;
+use crate::info::Info;
+
+const RAH_TYPE_ID: i32 = 4403;
+const RESISTANCE_SHIFT_AMOUNT_ID: i32 = 1849;
+const RESONANCE_IDS: [i32; 4] = [267, 268, 269, 270]; // Order: Em, Explo, Kin, Therm
+
+pub fn adapt_rah_to_damage_pattern(info: &impl Info, ship: &mut Ship) {
+    let rah_index = if let Some(i) = ship
+        .items
+        .iter_mut()
+        .position(|item| item.type_id == RAH_TYPE_ID)
+    {
+        i
+    } else {
+        return;
+    };
+
+    let effective_resonance_ids = [
+        info.attribute_name_to_id("armorEmDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorExplosiveDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorKineticDamageEffectiveResonance"),
+        info.attribute_name_to_id("armorThermalDamageEffectiveResonance"),
+    ];
+
+    let rah_affected_attributes = rah_affected_attributes(ship, rah_index);
+    let shift_amount = ship.items[rah_index]
+        .attributes
+        .get(&RESISTANCE_SHIFT_AMOUNT_ID)
+        .and_then(|a| a.value)
+        .unwrap()
+        / 100.0;
+
+    let mut cache = Cache::default();
+    let mut cycle_list: Vec<[f64; 4]> = Vec::new();
+    let mut cycle_start = None;
+
+    'sim: for _ in 0..50 {
+        let effective_resonance =
+            get_effective_resonance(ship, &mut cache, effective_resonance_ids);
+        let rah_resonance = get_rah_resonance(ship, &mut cache, rah_index);
+
+        let shift = calculate_rah_shift(effective_resonance, rah_resonance, shift_amount);
+        let shifted = std::array::from_fn(|i| rah_resonance[i] + shift[i]);
+
+        for (i, val) in cycle_list.iter().enumerate() {
```

Comment:
Please add a bit of comments what is going on here. It is not a trivial for-body to read, so a comment is well deserved :)